File: filly-core/src/widgets/summary.rs
use crate::event::{Event, KeyCode};
use crate::render::{
    Alignment, BorderStyle, EdgeInsets, Rect, RenderNode, RenderTree, TextStyle,
};
use crate::widget::{EventResult, Widget};
use filly_protocol::WidgetResponse;
use std::fs;

pub struct SummaryWidget {
    pub title: String,
    pub text: String,
    pub scroll: usize,
    dirty: bool,
}

impl SummaryWidget {
    pub fn new(title: String, message: Option<String>, file: Option<String>) -> Self {
        let text = if let Some(path) = file {
            fs::read_to_string(&path).unwrap_or_else(|_| format!("[Error reading {}]", path))
        } else {
            message.unwrap_or_default()
        };
        Self {
            title,
            text,
            scroll: 0,
            dirty: true,
        }
    }
}

impl Widget for SummaryWidget {
    fn render(&self, area: Rect) -> RenderTree {
        let box_w = ((area.width as f32 * 0.8) as u16).min(area.width.saturating_sub(2));
        let box_h = ((area.height as f32 * 0.8) as u16).min(area.height.saturating_sub(2));
        let box_x = (area.width.saturating_sub(box_w)) / 2;
        let box_y = (area.height.saturating_sub(box_h)) / 2;


        let mut children = Vec::new();
        children.push(RenderNode::Text {
            rect: Rect::new(1, 0, box_w - 2, 1),
            content: self.title.clone(),
            alignment: Alignment::Center,
            role: None, label: None, description: None, style: TextStyle {
                bold: true,
                fg: Some("green".into()),
                ..TextStyle::normal()
            },
        });
        children.push(RenderNode::Text {
            rect: Rect::new(1, 1, box_w - 2, box_h - 2),
            content: self.text.clone(),
            alignment: Alignment::Left,
            style: TextStyle::normal(), role: None, label: None, description: None,
        });
        children.push(RenderNode::Text {
            rect: Rect::new(1, box_h - 1, box_w - 2, 1),
            content: "Any key to continue  Esc:cancel".into(),
            alignment: Alignment::Center,
            style: TextStyle::muted(), role: None, label: None, description: None,
        });

        RenderTree::Container {
            rect: Rect::new(box_x, box_y, box_w, box_h),
            background: None,
            border: BorderStyle::Single,
            padding: EdgeInsets::zero(), role: None, label: None, description: None,
            children,
        }
    }

    fn handle_event(&mut self, event: &Event) -> EventResult {
        if let Event::Key(code, _) = event {
            match code {
                KeyCode::Esc => EventResult::Response(WidgetResponse {
                    result: None,
                    cancelled: true,
                    error: None,
                }),
                _ => EventResult::Response(WidgetResponse {
                    result: None,
                    cancelled: false,
                    error: None,
                }),
            }
        } else {
            EventResult::Unhandled
        }
    }

    fn is_dirty(&self) -> bool {
        self.dirty
    }
    fn clear_dirty(&mut self) {
        self.dirty = false;
    }
}

File: filly-core/widgets/summary.c
#include "summary.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

typedef struct {
    char *title;
    char *text;
    int scroll;
    bool dirty;
} SummaryData;

static void summary_render(Widget *self, Rect area, RenderTree *out) {
    SummaryData *d = (SummaryData *)(self + 1);
    memset(out, 0, sizeof(*out));
    int box_w = (int)(area.w * 0.8f);
    if (box_w > area.w - 2) box_w = area.w - 2;
    int box_h = (int)(area.h * 0.8f);
    if (box_h > area.h - 2) box_h = area.h - 2;
    int box_x = (area.w - box_w) / 2, box_y = (area.h - box_h) / 2;

    out->accessible.role = strdup("dialog");
    out->accessible.label = strdup(d->title ? d->title : "Summary");

    RenderTree *children = calloc(3 + 1, sizeof(RenderTree));
    children[0].type = RNODE_TEXT;
    children[0].rect = rect_new(1, 0, box_w - 2, 1);
    children[0].text.content = strdup(d->title);
    children[0].text.align = ALIGN_CENTER;
    children[0].text.style = textstyle_selected();
    children[0].accessible.role = strdup("heading");
    children[0].accessible.label = strdup(d->title);

    children[1].type = RNODE_TEXT;
    children[1].rect = rect_new(1, 1, box_w - 2, box_h - 3);
    children[1].text.content = strdup(d->text);
    children[1].text.align = ALIGN_LEFT;
    children[1].text.style = textstyle_normal();

    children[2].type = RNODE_TEXT;
    children[2].rect = rect_new(1, box_h - 2, box_w - 2, 1);
    children[2].text.content = strdup("Any key to continue  Esc:cancel");
    children[2].text.align = ALIGN_CENTER;
    children[2].text.style = textstyle_muted();

    out->type = RNODE_CONTAINER;
    out->rect = rect_new(box_x, box_y, box_w, box_h);
    out->container.border = BORDER_SINGLE;
    out->container.padding = edgeinsets_zero();
    out->container.children = children;
    out->container.child_count = 3;
}

static EventResult summary_handle_event(Widget *self, Event *ev, Backend *backend) {
    (void)self; (void)backend;
    if (ev->type == EVENT_KEY) {
        if (ev->code == KEY_ESC)
            return event_result_response((WidgetResponse){ .result = NULL, .cancelled = true, .error = NULL });
        return event_result_response((WidgetResponse){ .result = NULL, .cancelled = false, .error = NULL });
    }
    return event_result_unhandled();
}

static bool summary_is_dirty(Widget *self) { return ((SummaryData *)(self + 1))->dirty; }
static void summary_clear_dirty(Widget *self) { ((SummaryData *)(self + 1))->dirty = false; }
static void summary_destroy(Widget *self) {
    SummaryData *d = (SummaryData *)(self + 1);
    free(d->title); free(d->text);
}

Widget *summary_widget_new(const char *title, const char *message, const char *file_path) {
    Widget *w = calloc(1, sizeof(Widget) + sizeof(SummaryData));
    w->vtable.render = summary_render;
    w->vtable.handle_event = summary_handle_event;
    w->vtable.is_dirty = summary_is_dirty;
    w->vtable.clear_dirty = summary_clear_dirty;
    w->vtable.destroy = summary_destroy;
    SummaryData *d = (SummaryData *)(w + 1);
    d->title = strdup(title);
    if (file_path) {
        FILE *f = fopen(file_path, "r");
        if (f) { fseek(f, 0, SEEK_END); long sz = ftell(f); rewind(f); d->text = malloc(sz + 1); fread(d->text, 1, sz, f); d->text[sz] = '\0'; fclose(f); }
        else d->text = strdup("[Error reading file]");
    } else d->text = message ? strdup(message) : strdup("");
    d->scroll = 0; d->dirty = true;
    return w;
}